driver.go

```diff
@@ -327,40 +348,95 @@ func (d *Driver) StartTask(cfg *drivers.TaskConfig) (*drivers.TaskHandle, *drive
 	if driverConfig.Command != "" {
-		allArgs = append(allArgs, driverConfig.Command)
+		command = []string{driverConfig.Command}
 	}
+	allArgs = append(allArgs, command...)
 	allArgs = append(allArgs, driverConfig.Args...)
+
+	entryPoint := strings.Join(img.Config.Entrypoint, " ")
+	if driverConfig.Entrypoint != "" {
+		entryPoint = driverConfig.Entrypoint
+	}
+
+	workingDir := img.Config.WorkingDir
+	if workingDir == "" {
+		workingDir = "/"
+	}
+	if driverConfig.WorkingDir != "" {
+		workingDir = driverConfig.WorkingDir
+	}
+
 	containerName := BuildContainerName(cfg)
 	memoryLimit := fmt.Sprintf("%dm", cfg.Resources.NomadResources.Memory.MemoryMB)
 	cpuShares := cfg.Resources.LinuxResources.CPUShares
 	logOpts := []string{
 		fmt.Sprintf("path=%s", cfg.StdoutPath),
 	}
 
-	// ensure to include port_map into tasks environment map
-	cfg.Env = taskenv.SetPortMapEnvs(cfg.Env, driverConfig.PortMap)
-
-	// convert environment map into a k=v list
-	allEnv := cfg.EnvList()
+	// Handle mounting directories local to the allocation
+	// and arbitrary host paths if the driver has "volumes" enabled
+	var allocVolumes []string
+	volumeEnv := make(map[string]string)
+	for _, volume := range driverConfig.Volumes {
+		v := ""
+		if strings.HasPrefix(volume, "local/") {
+			v = strings.TrimPrefix(volume, "local/")
+			v = strings.Join([]string{cfg.TaskDir().LocalDir, v}, "/")
+			volumeEnv[taskenv.TaskLocalDir] = strings.Split(v,":")[1]
+		} else if strings.HasPrefix(volume, "alloc/") {
+			v = strings.TrimPrefix(volume, "alloc/")
+			v = strings.Join([]string{cfg.TaskDir().SharedAllocDir, v}, "/")
+			volumeEnv[taskenv.AllocDir] = strings.Split(v,":")[1]
+		} else if strings.HasPrefix(volume, "secrets/") {
+			v = strings.TrimPrefix(volume, "secrets/")
+			v = strings.Join([]string{cfg.TaskDir().SecretsDir, v}, "/")
+			volumeEnv[taskenv.SecretsDir] = strings.Split(v,":")[1]
+		} else {
+			// If the volume isn't relative to the allocations directory
+			// we need to be sure that mounting volumes from outside the
+			// allocation directory is allowed by the driver configuration.
+			if d.config.Volumes.Enabled {
+				v = volume
+			} else {
+				d.logger.Warn("Volumes", fmt.Sprintf("trying to mount %#v, which is outside the allocation directory, while volume mounting from host paths haven't been enabled", volume))
+			}
+		}
 
-	// ensure to mount nomad alloc dirs into the container
-	allVolumes := []string{
-		fmt.Sprintf("%s:%s", cfg.TaskDir().SharedAllocDir, cfg.Env[taskenv.AllocDir]),
-		fmt.Sprintf("%s:%s", cfg.TaskDir().LocalDir, cfg.Env[taskenv.TaskLocalDir]),
-		fmt.Sprintf("%s:%s", cfg.TaskDir().SecretsDir, cfg.Env[taskenv.SecretsDir]),
+		allocVolumes = append(allocVolumes, v)
 	}
 
-	if d.config.Volumes.Enabled {
-		// add task specific volumes, if enabled
-		allVolumes = append(allVolumes, driverConfig.Volumes...)
+	d.logger.Debug("Volumes", fmt.Sprintf("%#v", allocVolumes))
+
+	// reset env so we don't inherit host env by default
```

Comment:
Tested this out and compared it to the Docker driver, and yeah this isn't correct. From my current testing it looks like the right way would be to reset the env to only contain variables prefixed with `NOMAD_*`, then overlay the image defaults on top of those, and lastly pick any task config env variables (which I don't think are implemented yet).

What comes to the volumes I think there's a small difference in how Docker and Podman work here, if we look at how Docker driver handles this: https://github.com/hashicorp/nomad/blob/master/drivers/docker/driver.go#L635-L687 where `docker inspect` shows bind mounts under HostConfig like this:

```
        "HostConfig": {
            "Binds": [
                "/<path-to-nomad-data-dir>/alloc/fd74a661-0766-9c4a-7224-77769b92a0b6/alloc:/alloc:z",
        ...
```

and then separate mounts from that `/alloc` to what's defined in the task config in `volumes = [...]`. It doesn't look like that's needed in Podman, so my reasoning for this was we mount the `/local`, `/secrets` and `/alloc` if it's requested and set the env variables to reflect where they were mounted. Maybe it doesn't make sense this way.

Reply:
I am not sure about the correct overal env handling. It will be best to learn from other drivers. Rawexec, in example, simply calls cfg.EnvList()  and so does the java and qemu drivers. Hence i added the same snipped to the podman driver.

Related to automatic alloc mounts: AFAIK the container should solely rely on env variables to discover Nomads goodies. It would be fine for me mounting it statically on /alloc etc as long as it's correctly reflected in the env.

Reply:
Cool, so that's what I went on and did now after toying around with this for a while :). So the weird env variables I get is actually from running things rootless, it's deriving my users environment which ofc makes it differ from the Docker ones. IMO no need to clear the env now that I realized that's what's happening. The only thing is that the env variables from the image will override the base env variables, which seems to do what's expected for me.

Now the main difference is to handle the task configured volumes like Docker driver does, so e.g. mounting `local/:/conf` gives you a directory within the container where `<nomad-data-dir>/<alloc-local-dir>` is mounted to `/conf` inside the container.

The default alloc mounts are now mounted to `/alloc`, `/local` and `/secrets` and env variables are set to those paths, this is what I noticed Docker driver is doing.

Reply:
It's much better now, thank you!